resolwe/flow/executors/docker/run.py

```diff
@@ -115,26 +152,32 @@ def start(self):
         group_file.file.flush()
         self.temporary_files.append(group_file)
 
-        mappings.append({'src': passwd_file.name, 'dest': '/etc/passwd', 'mode': 'ro,Z'})
-        mappings.append({'src': group_file.name, 'dest': '/etc/group', 'mode': 'ro,Z'})
+        add_volume(volumes, 'users', None, '/etc/passwd', [passwd_file.name])
+        add_volume(volumes, 'users', None, '/etc/group', [group_file.name])
 
-        # create mappings for tools
+        # Create volumes for tools.
         # NOTE: To prevent processes tampering with tools, all tools are mounted read-only
-        # NOTE: Since the tools are shared among all containers they must use the shared SELinux
-        # label (z option)
-        self.mappings_tools = [{'src': tool, 'dest': '/usr/local/bin/resolwe/{}'.format(i), 'mode': 'ro,z'}
-                               for i, tool in enumerate(self.get_tools())]
-        mappings += self.mappings_tools
-        # create Docker --volume parameters from mappings
-        command_args['volumes'] = ' '.join(['--volume="{src}":"{dest}":{mode}'.format(**map_)
-                                            for map_ in mappings])
-
-        # set working directory inside the container to the mapped directory of
-        # the current Data's directory
-        command_args['workdir'] = ''
-        for template in mappings_template:
-            if '{data_id}' in template['src']:
-                command_args['workdir'] = '--workdir={}'.format(template['dest'])
+        self.mappings_tools = []
+        for index, tool in enumerate(self.get_tools()):
+            add_volume(
+                self.mappings_tools,
+                'tools',
+                None,
+                os.path.join('/usr/local/bin/resolwe', str(index)),
+                [tool]
+            )
+
+        volumes += self.mappings_tools
+
+        # Add any extra volumes verbatim.
+        volumes += SETTINGS.get('FLOW_DOCKER_EXTRA_VOLUMES', [])
+
+        # Create Docker --volume parameters from volumes.
+        command_args['volumes'] = ' '.join(['--volume="{src}":"{dest}":{options}'.format(**volume)
```

Comment:
@kostko Will this work if `options` is empty? Or do you have to omit the colon?

Reply:
Yes, it works if the options are empty (no need to omit the colon).